swift/ql/src/queries/Security/CWE-312/CleartextStoragePreferences.ql

```diff
@@ -0,0 +1,105 @@
+/**
+ * @name Cleartext storage of sensitive information in application preferences
+ * @description Storing sensitive information in a non-encrypted database can expose it to an attacker.
+ * @kind path-problem
+ * @problem.severity warning
+ * @security-severity 7.5
+ * @precision medium
+ * @id swift/cleartext-storage-preferences
+ * @tags security
+ *       external/cwe/cwe-312
+ */
+
+import swift
+import codeql.swift.security.SensitiveExprs
+import codeql.swift.dataflow.DataFlow
+import codeql.swift.dataflow.TaintTracking
+import DataFlow::PathGraph
+
+/**
+ * A `DataFlow::Node` of something that gets stored in a preferences store.
+ */
+abstract class Stored extends DataFlow::Node {
+  abstract string getStoreName();
+}
+
+/** The `DataFlow::Node` of an expression that gets written to the defaults database */
+class UserDefaultsStore extends Stored {
+  UserDefaultsStore() {
+    exists(ClassDecl c, AbstractFunctionDecl f, CallExpr call |
+      c.getName() = ["UserDefaults"] and
+      c.getAMember() = f and
+      f.getName() = ["set(_:forKey:)"] and
+      call.getStaticTarget() = f and
+      call.getArgument(0).getExpr() = this.asExpr()
+    )
+  }
+
+  override string getStoreName() { result = "the user defaults database" }
+}
+
+/** The `DataFlow::Node` of an expression that gets written to iCloud */
```

Comment:
This is a great addition.  Do you think calling this "application preferences" in the query `@name` and the "defaults database" in the `.qhelp` is accurate enough when we're also including iCloud ... or should we look for a more generic term?

Reply:
I was trying to be generic, but it's kinda hard. Maybe: "an application preference store"? "Local and/or {shared,distributed} preferences"?

Reply:
Yeah I can see its difficult to be generic, while also being clear what we're talking about.  I think I prefer "an application preference store", because its readable but does imply more than one type of store is supported.  Whatever we pick we should try to be consistent.